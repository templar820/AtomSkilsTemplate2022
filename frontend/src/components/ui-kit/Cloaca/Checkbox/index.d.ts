import * as React from 'react';
import { EmotionStylesType } from 'lib-ui';

export interface CheckboxProps {
    /** Toggle control mode */
    isControlled?: boolean;
    /** Defines whether the checkbox is checked */
    checked?: boolean | 'indeterminate';
    /** Defines name of the checkbox */
    name?: string;
    /** Defines text or any node that will be placed in a body of the component */
    children?: React.ReactNode;
    /** Defines font-size */
    size?: string | number;
    /** Callback that fires when checkbox's state has changed, func(e) */
    onCheckChange?: (event: React.MouseEvent) => void;
    /** Defines whether checkbox is disabled */
    disabled?: boolean;
    /** Defines additional checkbox css */
    css?: EmotionStylesType;
    /** Defines additional frame css */
    squareStyle?: EmotionStylesType;
}

declare const Checkbox: React.ComponentType<CheckboxProps>;

export default Checkbox;
