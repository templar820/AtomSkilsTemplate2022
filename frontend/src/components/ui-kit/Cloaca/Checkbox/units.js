import React from 'react';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { autoAddPx } from 'lib-root/utils/styleMixins';
import { px } from 'lib-root/utils/units';
import applyEmotionStyle from 'lib-root/utils/applyEmotionStyle';

const StyledLabel = styled.label`
    display: inline-flex;
    align-items: center;
    margin: 0;
    ${({ isChildrenExist, size }) => {
        return (
            isChildrenExist &&
            css`
                padding-right: ${px(size, 0.375)};
            `
        );
    }};
    font-size: ${({ size }) => autoAddPx(size)};
    cursor: pointer;
    user-select: none;
    ${({ disabled }) => {
        return (
            disabled &&
            css`
                opacity: 0.25;
                cursor: not-allowed;
            `
        );
    }}

    ${({ _css, ...restProps }) => applyEmotionStyle(_css, restProps)};
`;

const CheckboxSquare = styled.span`
    ${({ colors, isChildrenExist, disabled, isHovered, _checked, squareStyle, size, ...restProps }) => css`
        position: relative;
        display: inline-block;
        width: 20px;
        min-width: 20px;
        height: 20px;
        min-height: 20px;
        margin: ${px(size, 0.375)};
        ${isChildrenExist &&
            css`
                margin-right: ${px(size, 0.875)};
            `};
        border-radius: ${px(size, 0.125)};
        background-color: ${disabled ? colors.GrayScale_200 : 'transparent'};
        color: ${isHovered ? colors.primaryAccent : colors.primary};
        border: ${px(size, 0.125)} solid
            ${disabled ? colors.GrayScale_200 : isHovered ? 'currentColor' : colors.GrayScale_200};
        .hover & {
            border-color: ${colors.primaryAccent};
            color: ${colors.primaryAccent};
        }
        svg {
            width: ${px(size, 1.25)};
            height: ${px(size, 1.25)};
            position: absolute;
            top: 50%;
            left: 50%;
            transform: translate(-50%, -50%);
        }
        ${_checked === 'indeterminate' &&
            css`
                &:after {
                    content: '';
                    position: absolute;
                    top: 50%;
                    left: 50%;
                    transform: translate(-50%, -50%);
                    width: ${px(size, 0.5)};
                    height: ${px(size, 0.5)};
                    border-radius: ${px(size, 0.0625)};
                    background-color: ${disabled ? colors.primary : 'currentColor'};
                }
            `};

        ${applyEmotionStyle(squareStyle, {
            colors,
            isChildrenExist,
            disabled,
            isHovered,
            _checked,
            size,
            ...restProps
        })};
    `}
`;

const CheckboxCaption = styled.span`
    overflow: hidden;
    font-size: 15px;
    line-height: 24px;
    ${({ captionStyle }) => captionStyle}
`;

const CheckboxSvg = ({ colors, disabled, _checked }) =>
    _checked === true && (
        <svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
            <path
                d="M8.83178 11.5475L6.61877 9.33452L5.20483 10.7485L7.66627 13.2102L9.11181 14.656L15.044 6.16961L13.4057 5.02246L8.83178 11.5475Z"
                fill={disabled ? colors.primary : 'currentColor'}
            />
        </svg>
    );

export { StyledLabel, CheckboxSquare, CheckboxCaption, CheckboxSvg };
