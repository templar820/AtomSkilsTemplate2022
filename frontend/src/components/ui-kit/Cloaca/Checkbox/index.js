import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ColorsContext } from 'lib-root/colors';
import { CheckboxSvg, CheckboxSquare, CheckboxCaption, StyledLabel } from './units';
import { setClassName } from 'lib-root/utils/styleMixins';

/**
 * Используется для управления параметром выбора с тремя состояниями **checked** – `true`, `false` и `indeterminate`.
 */
class Checkbox extends Component {
    inputRef = React.createRef();
    state = { isHovered: false, isChecked: this.props.checked };

    changeCheck(e) {
        const { onCheckChange, isControlled } = this.props;
        if (onCheckChange) onCheckChange(e);

        if (!isControlled) {
            this.setState((state) => {
                return {
                    isChecked: !state.isChecked
                };
            });
        }
    }

    setIndeterminateStatus() {
        this.inputRef.current.indeterminate = this.props.checked === 'indeterminate';
    }

    componentDidMount() {
        this.setIndeterminateStatus();
    }

    componentDidUpdate(prevProps) {
        if (this.props.checked !== prevProps.checked) this.setIndeterminateStatus();
    }

    render() {
        const {
            disabled,
            name,
            checked,
            isControlled,
            children,
            children: isChildrenExist,
            css: _css,
            squareStyle,
            captionStyle,
            size,
            className,
            onCheckChange,
            ...restProps
        } = this.props;
        const { isHovered, isChecked } = this.state;

        const _checked = isControlled ? checked : isChecked;

        return (
            <ColorsContext.Consumer>
                {(colors) => (
                    <StyledLabel
                        onMouseEnter={() => this.setState({ isHovered: true })}
                        onMouseLeave={() => this.setState({ isHovered: false })}
                        {...{ disabled, _css, size, isChildrenExist }}
                        className={setClassName({ props: this.props, name: 'checkbox' })}>
                        <input
                            ref={this.inputRef}
                            type="checkbox"
                            hidden={true}
                            onChange={(e) => this.changeCheck(e)}
                            checked={_checked === true}
                            {...{ disabled, name, ...restProps }}
                        />
                        <CheckboxSquare
                            {...{ colors, disabled, isHovered, _checked, isChildrenExist, squareStyle, size }}>
                            <CheckboxSvg {...{ colors, _checked, disabled }} />
                        </CheckboxSquare>
                        <CheckboxCaption {...{ captionStyle }}>{children}</CheckboxCaption>
                    </StyledLabel>
                )}
            </ColorsContext.Consumer>
        );
    }
}
Checkbox.displayName = 'Checkbox';
Checkbox.propTypes = {
    /** Toggle control mode */
    isControlled: PropTypes.bool,
    /** Defines whether the checkbox is checked */
    checked: PropTypes.oneOf([true, false, 'indeterminate']),
    /** Defines name of the checkbox */
    name: PropTypes.string,
    /** Defines text or any node that will be placed in a body of the component */
    children: PropTypes.node,
    /** Defines font-size */
    size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** Callback that fires when checkbox's state has changed, func(e) */
    onCheckChange: PropTypes.func,
    /** Defines whether checkbox is disabled */
    disabled: PropTypes.bool,
    /** Defines additional checkbox css */
    css: PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.string]),
    /** Defines additional frame css */
    squareStyle: PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.string])
};
Checkbox.defaultProps = {
    isControlled: true,
    checked: false,
    disabled: false,
    size: '16px'
};

export default Checkbox;
