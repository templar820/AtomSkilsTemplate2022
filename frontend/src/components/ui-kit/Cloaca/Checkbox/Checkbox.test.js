import React from 'react';
import ReactDOM from 'react-dom';
// needs for serialize emotion classes to the snapshot
import serializer from 'jest-emotion';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import Checkbox from './';

expect.addSnapshotSerializer(serializer);

Enzyme.configure({ adapter: new Adapter() });

const myMockFn = jest.fn();

const props = {
    checked: 'indeterminate',
    children: <div>Checkbox label</div>,
    name: 'input-name',
    size: 16,
    className: 'some-class',
    onCheckChange: myMockFn,
    disabled: true,
    css: { fontSize: '18px' },
    squareStyle: { width: '18px' }
};

describe('<Checkbox />', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = mount(<Checkbox {...props} />);
    });

    afterEach(() => {
        wrapper.unmount();
    });

    // test must be first, otherwise the snapshot size will be too large
    test('should match snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Checkbox />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('should render correctly', () => {
        expect(wrapper.find('input').length).toBe(1);

        expect(wrapper.find('input')).toBeDisabled();
        wrapper.setProps({ disabled: false });
        expect(wrapper.find('input')).not.toBeDisabled();

        expect(wrapper.find('input')).not.toBeChecked();
        wrapper.setProps({ checked: true });
        expect(wrapper.find('input')).toBeChecked();
    });

    test('should change state on mouse events', () => {
        expect(wrapper.state().isHovered).toBe(false);
        wrapper.find('label').simulate('mouseEnter');
        expect(wrapper.state().isHovered).toBe(true);
        wrapper.find('label').simulate('mouseLeave');
        expect(wrapper.state().isHovered).toBe(false);
    });

    test('should call onCheckChange method', () => {
        wrapper.find('input').simulate('change', { target: { value: true } });

        expect(myMockFn).toHaveBeenCalled();
    });
});
