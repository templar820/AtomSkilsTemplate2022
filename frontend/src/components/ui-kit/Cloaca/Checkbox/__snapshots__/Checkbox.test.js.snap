// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should match snapshot 1`] = `
.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0;
  padding-right: 6px;
  font-size: 16px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  opacity: 0.25;
  cursor: not-allowed;
  font-size: 18px;
}

.emotion-0 {
  position: relative;
  display: inline-block;
  width: 20px;
  min-width: 20px;
  height: 20px;
  min-height: 20px;
  margin: 6px;
  margin-right: 14px;
  border-radius: 2px;
  background-color: #D8D8D8;
  color: #00B956;
  border: 2px solid #D8D8D8;
  width: 18px;
}

.hover .emotion-0 {
  border-color: #10E272;
  color: #10E272;
}

.emotion-0 svg {
  width: 20px;
  height: 20px;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.emotion-0:after {
  content: '';
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  width: 8px;
  height: 8px;
  border-radius: 1px;
  background-color: #00B956;
}

.emotion-2 {
  overflow: hidden;
  font-size: 15px;
  line-height: 24px;
}

<Checkbox
  checked="indeterminate"
  className="some-class"
  css={
    Object {
      "fontSize": "18px",
    }
  }
  disabled={true}
  isControlled={true}
  name="input-name"
  onCheckChange={[MockFunction]}
  size={16}
  squareStyle={
    Object {
      "width": "18px",
    }
  }
>
  <StyledLabel
    _css={
      Object {
        "fontSize": "18px",
      }
    }
    className="some-class CLR-checkbox"
    disabled={true}
    isChildrenExist={
      <div>
        Checkbox label
      </div>
    }
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size={16}
  >
    <label
      className="some-class CLR-checkbox emotion-4 emotion-5"
      disabled={true}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      size={16}
    >
      <input
        checked={false}
        disabled={true}
        hidden={true}
        name="input-name"
        onChange={[Function]}
        type="checkbox"
      />
      <CheckboxSquare
        _checked="indeterminate"
        colors={
          Object {
            "GrayScale_0": "#ffffff",
            "GrayScale_100": "#EDEDED",
            "GrayScale_150": "#E5E5E5",
            "GrayScale_200": "#D8D8D8",
            "GrayScale_50": "#fafafa",
            "GrayScale_500": "#999999",
            "GrayScale_55": "#f6f6f6",
            "GrayScale_600": "#7C7C7C",
            "GrayScale_700": "#333333",
            "GrayScale_900": "#000000",
            "error": "#F62434",
            "errorAccent": "#EB5A40",
            "info": "#34AAF2",
            "infoAccent": "#5BD9E5",
            "primary": "#00B956",
            "primaryAccent": "#10E272",
            "secondary": "#731982",
            "secondaryAccent": "#A500BF",
            "tertiary": "#444189",
            "tertiaryAccent": "#404D46",
            "warning": "#FFA717",
            "warningAccent": "#F36F20",
          }
        }
        disabled={true}
        isChildrenExist={
          <div>
            Checkbox label
          </div>
        }
        isHovered={false}
        size={16}
        squareStyle={
          Object {
            "width": "18px",
          }
        }
      >
        <span
          className="emotion-0 emotion-1"
          disabled={true}
          size={16}
        >
          <CheckboxSvg
            _checked="indeterminate"
            colors={
              Object {
                "GrayScale_0": "#ffffff",
                "GrayScale_100": "#EDEDED",
                "GrayScale_150": "#E5E5E5",
                "GrayScale_200": "#D8D8D8",
                "GrayScale_50": "#fafafa",
                "GrayScale_500": "#999999",
                "GrayScale_55": "#f6f6f6",
                "GrayScale_600": "#7C7C7C",
                "GrayScale_700": "#333333",
                "GrayScale_900": "#000000",
                "error": "#F62434",
                "errorAccent": "#EB5A40",
                "info": "#34AAF2",
                "infoAccent": "#5BD9E5",
                "primary": "#00B956",
                "primaryAccent": "#10E272",
                "secondary": "#731982",
                "secondaryAccent": "#A500BF",
                "tertiary": "#444189",
                "tertiaryAccent": "#404D46",
                "warning": "#FFA717",
                "warningAccent": "#F36F20",
              }
            }
            disabled={true}
          />
        </span>
      </CheckboxSquare>
      <CheckboxCaption>
        <span
          className="emotion-2 emotion-3"
        >
          <div>
            Checkbox label
          </div>
        </span>
      </CheckboxCaption>
    </label>
  </StyledLabel>
</Checkbox>
`;
